### PR TITLE
Preserve tree attachment in the typer's Block/Apply inversion transform

### PR DESCRIPTION
**Pull request by retronym**
This keeps NamedApplyInfo attachment in place for downstream analysis
by compiler plugins.

Fixes scala/bug#12659

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -4919,8 +4919,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       }
 
       def typedApply(tree: Apply) = tree match {
-        case Apply(Block(stats, expr), args) =>
-          typed1(atPos(tree.pos)(Block(stats, Apply(expr, args) setPos tree.pos.makeTransparent)), mode, pt)
+        case Apply(blk @ Block(stats, expr), args) =>
+          val ap1 = treeCopy.Apply(tree, expr, args).clearType().setPos(tree.pos.makeTransparent)
+          val blk1 = treeCopy.Block(blk, stats, ap1).clearType()
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:4924]: Why did you add `clearType`?

Comment (retronym) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:4924]: I wanted to pass on the attachment but I didn't want to change the current approach of re-typechecking the new `Apply` and `Block`

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:4924]: Ah of course... Thanks

**Review by lrytz — APPROVED** [review]
